src/adapter/src/coord/statement_logging.rs

```diff
@@ -43,13 +46,116 @@ pub enum PreparedStatementLoggingInfo {
+        let mut out = None;
+
+        let uuid = match logging {
+            PreparedStatementLoggingInfo::AlreadyLogged { uuid } => *uuid,
+            PreparedStatementLoggingInfo::StillToLog {
+                sql,
+                prepared_at,
+                name,
+                session_id,
+                accounted,
+            } => {
+                assert!(
+                    *accounted,
+                    "accounting for logging should be done in `begin_statement_execution`"
+                );
+                let uuid = Uuid::new_v4();
+                out = Some(StatementPreparedRecord {
+                    id: uuid,
+                    sql: std::mem::take(sql),
+                    name: std::mem::take(name),
+                    session_id: *session_id,
+                    prepared_at: *prepared_at,
+                });
+
+                *logging = PreparedStatementLoggingInfo::AlreadyLogged { uuid };
+                uuid
+            }
+        };
+        (out, uuid)
+    }
+    /// The rate at which statement execution should be sampled.
+    /// This is the value of the session var `statement_logging_sample_rate`,
+    /// constrained by the system var `statement_logging_max_sample_rate`.
+    pub fn statement_execution_sample_rate(&self, session: &Session) -> f64 {
+        let system: f64 = self
+            .catalog()
+            .system_config()
+            .statement_logging_max_sample_rate()
+            .try_into()
+            .expect("value constrained to be convertible to f64");
+        let user: f64 = session
+            .vars()
+            .get_statement_logging_sample_rate()
+            .try_into()
+            .expect("value constrained to be convertible to f64");
+        f64::min(system, user)
+    }
+
+    /// Record the end of statement execution for a statement whose beginning was logged.
+    /// It is an error to call this function for a statement whose beginning was not logged
+    /// (because it was not sampled). Requiring the opaque `StatementLoggingId` type,
+    /// which is only instantiated by `begin_statement_execution` if the statement is actually logged,
+    /// should prevent this.
+    pub fn end_statement_execution(
+        &mut self,
+        StatementLoggingId(id): StatementLoggingId,
+        reason: StatementEndedExecutionReason,
     ) {
+        let now = self.now_datetime();
+        let now_millis = now.timestamp_millis().try_into().expect("sane system time");
+        let ended_record = StatementEndedExecutionRecord {
+            id,
+            reason,
+            ended_at: now_millis,
+        };
+
+        let began_record = self.statement_logging_executions_begun.remove(&id).expect(
+            "matched `begin_statement_execution` and `end_statement_execution` invocations",
+        );
+        let updates = self
+            .catalog
+            .state()
+            .pack_statement_ended_execution_updates(&began_record, &ended_record);
```

Comment:
From a quick look it seems like the only part of this that needs the `Catalog` is a call to `Catalog::resolve_builtin_table(...)`. Can we do that call separately to decouple the rest of the packing process from the catalog? If need be this would allow us to move the work off the coord thread sometime in the future

Reply:
I could, but I'm a bit confused -- isn't having the `pack_xyz` methods be on `Catalog` (and internally call `resolve_builtin_table`) the standard pattern? Do we have plans to change that, or is there something specific about this case that means it makes sense to do it elsewhere?

Reply:
The coordinator can already hand out cheap Catalog references that can be moved off thread if needed. We don't need to change the pack pattern to accommodate.